tesseract_kinematics/src/core/kinematic_group.cpp

```diff
@@ -0,0 +1,135 @@
+    for (const auto& static_link_name : static_link_names_)
+      inv_working_frames_map_[static_link_name] = working_frame;
+  }
+  else
+  {
+    std::vector<std::string> child_link_names = scene_graph.getLinkChildrenNames(working_frame);
+    working_frames_.push_back(working_frame);
+    working_frames_.insert(working_frames_.end(), child_link_names.begin(), child_link_names.end());
+
+    inv_working_frames_map_[working_frame] = working_frame;
+    for (const auto& child : child_link_names)
+      inv_working_frames_map_[child] = working_frame;
+  }
+
+  tip_link_names_ = inv_kin_->getTipLinkNames();
+  for (const auto& tip_link : inv_kin_->getTipLinkNames())
+  {
+    std::vector<std::string> child_link_names = scene_graph.getLinkChildrenNames(tip_link);
+    tip_link_names_.insert(tip_link_names_.end(), child_link_names.begin(), child_link_names.end());
+
+    inv_tip_links_map_[tip_link] = tip_link;
+    for (const auto& child : child_link_names)
+      inv_tip_links_map_[child] = tip_link;
+  }
+
+  inv_to_fwd_base_ = state_.link_transforms.at(inv_kin_->getBaseLinkName()).inverse() *
+                     state_.link_transforms.at(state_solver_->getBaseLinkName());
+}
+
+KinematicGroup::KinematicGroup(const KinematicGroup& other) : JointGroup(other) { *this = other; }
+
+KinematicGroup& KinematicGroup::operator=(const KinematicGroup& other)
+{
+  JointGroup::operator=(other);
+  inv_kin_ = other.inv_kin_->clone();
+  inv_to_fwd_base_ = other.inv_to_fwd_base_;
+  working_frames_ = other.working_frames_;
+  tip_link_names_ = other.tip_link_names_;
+  inv_working_frames_map_ = other.inv_working_frames_map_;
+  inv_tip_links_map_ = other.inv_tip_links_map_;
+  return *this;
+}
+
+IKSolutions KinematicGroup::calcInvKin(const KinGroupIKInputs& tip_link_poses,
+                                       const Eigen::Ref<const Eigen::VectorXd>& seed) const
+{
+  // Convert to IK Inputs
+  IKInput ik_inputs;
+  for (const auto& tip_link_pose : tip_link_poses)
+  {
+    assert(std::find(tip_link_names_.begin(), tip_link_names_.end(), tip_link_pose.tip_link_name) !=
+           tip_link_names_.end());
+    assert(std::find(working_frames_.begin(), working_frames_.end(), tip_link_pose.working_frame) !=
+           working_frames_.end());
+
+    std::string inv_tip_link = inv_tip_links_map_.at(tip_link_pose.tip_link_name);
+    std::string inv_working_frame = inv_working_frames_map_.at(tip_link_pose.working_frame);
+
+    Eigen::Isometry3d wf_pose_offset =
+        state_.link_transforms.at(inv_working_frame).inverse() * state_.link_transforms.at(tip_link_pose.working_frame);
+    Eigen::Isometry3d tl_pose_offset =
+        state_.link_transforms.at(inv_tip_link).inverse() * state_.link_transforms.at(tip_link_pose.tip_link_name);
+
+    /** @todo Check this math (Levi) */
+    Eigen::Isometry3d inv_pose = wf_pose_offset * tip_link_pose.pose * tl_pose_offset.inverse();
+    ik_inputs[inv_tip_link] = inv_pose;
```

Comment:
Being extremely explicit and using descriptive naming usually helps ensure the math is correct. I would change to something like:

```c++
// Get transform from working frame to user working frame (reference frame for the target IK pose)
const Eigen::Isometry3d& world_to_user_wf = state_.link_transforms.at(tip_link_pose.working_frame);
const Eigen::Isometry3d& world_to_wf = state_.link_transforms.at(inv_kin_->getWorkingFrame());
const Eigen::Isometry3d wf_to_user_wf = world_to_wf.inverse() * world_to_user_wf;

// Get the transform from the user tip link to the IK solver tip link
const Eigen::Isometry3d& world_to_user_tl = ...
const Eigen::Isometry3d& world_to_tl = ...
const Eigen::Isometry3d user_tl_to_tl = world_to_user_tl.inverse() * world_to_tl;

// Get the transformation from the IK solver working frame to the IK solver tip frame
const Eigen::Isometry3d& user_wf_to_user_tl = tip_link_pose.pose;
const Eigen::Isometry3d wf_to_tl = wf_to_user_wf * user_wf_to_user_tl * user_tl_to_tl;
```

Reply:
Updated.